primitives/npos-elections/src/pjr.rs

```diff
@@ -0,0 +1,607 @@
+/// corresponds to `t == 0`.
+///
+/// However, standard PJR is less stringent than that. This function returns the threshold whose
+/// strength corresponds to the standard PJR property.
+///
+/// - `committed_size` is the number of winners of the election.
+/// - `weights` is an iterator of voter stakes. If the sum of stakes is already known,
+///   `std::iter::once(sum_of_stakes)` is appropriate here.
+pub fn standard_threshold(
+	committeed_size: usize,
+	weights: impl IntoIterator<Item = ExtendedBalance>,
+) -> Threshold {
+	weights
+		.into_iter()
+		.sum::<Threshold>()
+	/ committeed_size as Threshold
+}
+
+/// Check a solution to be PJR.
+///
+/// The PJR property is true if `t-PJR` is true when `t == sum(stake) / committee_size`.
+///
+/// This function returns an `AccountId` in the `Err` case. This is the counterexample: the ID of the
+/// unelected candidate with the highest prescore, such that `pre_score(counterexample) >= t`.
+pub fn pjr_check<AccountId: IdentifierT>(
+	supports: &Supports<AccountId>,
+	all_candidates: Vec<AccountId>,
+	all_voters: Vec<(AccountId, VoteWeight, Vec<AccountId>)>,
+) -> Result<(), AccountId> {
+	let t = standard_threshold(supports.len(), all_voters.iter().map(|voter| voter.1 as ExtendedBalance));
+	t_pjr_check(supports, all_candidates, all_voters, t)
+}
+
+/// Validate a challenge to an election result.
+///
+/// A challenge to an election result is valid if there exists some counterexample for which
+/// `pre_score(counterexample) >= threshold`. Validating an existing counterexample is computationally
+/// cheaper than re-running the PJR check.
+///
+/// Returns `true` if the challenge is valid: the proposed solution does not satisfy PJR.
+/// Returns `false` if the challenge is invalid: the proposed solution does in fact satisfy PJR.
+fn validate_pjr_challenge_core<AccountId: IdentifierT>(
+	counterexample: AccountId,
+	candidates: &[CandidatePtr<AccountId>],
+	voters: &[Voter<AccountId>],
+	threshold: Threshold,
+) -> bool {
+	let candidate = match candidates.iter().find(|candidate| candidate.borrow().who == counterexample) {
+		None => return false,
+		Some(candidate) => candidate.clone(),
+	};
+	pre_score(candidate, &voters, threshold) >= threshold
+}
+
+/// Validate a challenge to an election result.
+///
+/// A challenge to an election result is valid if there exists some counterexample for which
+/// `pre_score(counterexample) >= threshold`. Validating an existing counterexample is computationally
+/// cheaper than re-running the PJR check.
+///
+/// This function uses a supplied threshold.
+///
+/// Returns `true` if the challenge is valid: the proposed solution does not satisfy PJR.
+/// Returns `false` if the challenge is invalid: the proposed solution does in fact satisfy PJR.
+pub fn validate_t_pjr_challenge<AccountId: IdentifierT>(
+	counterexample: AccountId,
+	supports: &Supports<AccountId>,
+	all_candidates: Vec<AccountId>,
```

Comment:
I haven't read the whole code, but this is wrong IMO. The whole point of the challenge is to _not_ need to go over _all candidate_, and the challenge would just say which candidate we must look into exactly. Thus, you should not need to pass in all the candidates again. 

Voters are sadly unavoidable at first glance, but I guess we can improve that as well.

Reply:
I understand the motivation, but I think we can't avoid requiring those inputs. The elections code requires several complicated data structures: `Candidate`, `Voter`, `CandidatePtr`, which inter-relate to each other in complicated ways. If those complicated data structures were universal, then we could just have the calling code pass them in. As things stand, we have to construct them. 

"Why not add have the calling module cache the `Vec<Voter<AccountId>>` and restore it, instead of reconstructing it here?" Because `Voter` contains several `Edge`s, and `Edge` contains a `CandidatePtr` which is a `Rc<RefCell<Candidate<AccountId>>>>`, and ref-counted pointers are a pain to serialize and deserialize properly.